### DISCUSSION TITLE
Text display: Timing issue

### DISCUSSION THREAD

**Discussion by Duzeper** [pi3d/pi3d_demos#47] [Q&A]
I still have pictureFrame2020 working great with on/off timers and all the MQTT commands also working well.
I have, however, noticed something that is a little irritating. If I turn on or off any of the show text options, although they all work properly, the time delay is immediately reset to the default time, I presume by 'refresh = True'. Is there any way to make the image display just continue from where it was when the text display was initiated rather than starting again from zero?

Also, I have been looking at the discussion re. 'Current Date/Time' and saw your comment about adding a colon to CODEPOINTS in PictureFrame2020config.py.
You may remember I had an issue with the picture frame crashing when it saw an Icelandic character, which you fixed by trapping unexpected characters.
I wonder, though, given that 'NotoSans-Regular.ttf' supports the Icelandic alphabet, is it possible to simply add the character(s) to to 'CODEPOINTS  = ....'?
Assuming the answer is yes, then does it matter where in the string I insert the character?
Also, since CODEPOINTS is a quoted string, how would I also insert a single quote?

Finally, can you explain the CODEPOINTS = '... line?
I am assuming that the string is a list of characters that are available for the program to use, but what is the relevance of the comment 'limit to 49 ie 7x7 grid_size' given that there are more than 49 characters in the string?

**Comment by paddywwoof** [comment]
Hi, Sorry not to respond earlier. I get an email notification of Discussion posts just the same as Issues but I've been away from the laptop for a few days, checking emails with phone and, because I didn't reply when I read it, it's dropped off the radar!

The `refresh = True` behavior is a simple way to get the text to show as soon as you send an MQTT message turning it on. If the slide isn't refreshed then the the MQTT message will take immediate effect (as you might expect) however if the time through the slide is after the point where the text has faded it won't be visible. This is because the visibility of the text as well as the fading from one slide to the next depends on the progress of time https://github.com/pi3d/pi3d_demos/blob/master/PictureFrame2020.py#L528

In order for the text to start showing as soon as an MQTT message was received there would have to be an additional variable introduced into the above formula: `text_start_tm` with additional code to set and use it. Might not be too complicated, I will have a further look.

The CODEPOINTS line now looks like https://github.com/pi3d/pi3d_demos/blob/develop/PictureFrame2020config.py#L127 in the develop branch (which I will push to master quite soon, maybe after thinking about the `text_start_tm` idea). And here is a copy of my recent email correspondence with Wolfgang explaining it.

> Hi, Is your question following up on this recent discussion https://github.com/pi3d/pi3d_demos/discussions/36 ? The characters in CODEPOINTS are rendered onto a Texture 1024x1024 pixels so the more characters there are the smaller they need to be. The default string in config is 95 long so that's a 10x10 grid with each char in its own 102x102 pixel square. There's space left around the outside to allow for italic characters and rotated square not catching the corner of an adjacent char.
>
> So in conclusion: adding 5 more characters won't change anything and a few hundred more will probably be OK but if you had 1,600 characters each one would only have 25x25 pixels so you might start to see some graininess.
>
>And yes this would be something better explained rather than letting people find out by trial and error!
>
> Paddy

Paddy

**Reply by Duzeper** [reply]
Thanks. (been there done that)

I see that the new CODEPOINTS line is double quoted so that a single quote is now incorporated, but I still wonder, is that string a list of all the characters that the pictureFrame can use, and if so, can I simply add new characters to the string or, perhaps, replace characters that I am unlikely to need with those that I will?

Martin

**Comment by paddywwoof** [comment]
Hi, yes to all three questions. It's possible to read characters drawn on a 10x10 pixel square but that's probably the limit. i.e. in the 1024x1024 Texture there's space for up to 10,000 characters but at that point it looks pretty pixelly.
![10x100eg](https://user-images.githubusercontent.com/1939398/103645697-59f2e280-4f50-11eb-8b2e-4dd8fad0257d.png)
To add an apostrophe in a python string delimited by apostrophes you need to escape it. i.e.
```python
txt = '\'' #or "\""
```
I have modified the PictureFrame code in the develop branch https://github.com/pi3d/pi3d_demos/commit/bec6b0af2cbd183750037ee0f0e787cea74abab6 so it doesn't restart the slide when you change the text. As you can see it was a little more involved than I thought because the text creation code only ran at the start of a new slide.

Paddy

**Comment by jeffalperin** [comment]
A slightly different question about timing of text display:  Is it possible to have the text display begin either:
* immediately after **_completion_** of the slide transition; or
* after a user-configurable time delay after the **_beginning_** of the slide display
The problem I'm having is that the text display sometimes occurs in the midst of the slide transition making it difficult to determine whether the text describes the fading-out slide or the fading-in slide.  (I'm using a 4k display if that matters.)

--Jeff

**Reply by paddywwoof** [reply]
Yes, that does make sense really, and with the new system using a text_start_tm that could be accomplished very simply. I will try it on the frame here and implement it if it looks good.

Paddy

**Reply by paddywwoof** [reply]
Looks better. I've added that to the github develop branch
